src/debugging/DockerServerReadyAction.ts

```diff
@@ -213,54 +190,187 @@ export class ServerReadyDetector extends vscode.Disposable {
     }
 }
 
-export function registerServerReadyAction(context: vscode.ExtensionContext): void {
+type DebugAdapterMessage = {
+    body?: {
+        category?: string;
+        output?: string;
+    };
+    type?: string;
+    event?: string;
+};
+
+interface DockerServerReadyManager {
+    detectPattern(output: string): void;
+}
 
-    context.subscriptions.push(vscode.debug.onDidTerminateDebugSession(session => {
-        ServerReadyDetector.stop(session);
-    }));
+type LogStream = NodeJS.ReadableStream & { destroy(): void; };
+
+class DockerLogsTracker extends vscode.Disposable {
+    private logStream: LogStream;
 
-    const trackers = new Set<string>();
+    public constructor(containerName: string, detector: DockerServerReadyManager) {
+        super(
+            () => {
+                this.disposeStream();
+            });
 
-    context.subscriptions.push(vscode.debug.registerDebugConfigurationProvider('*', {
-        resolveDebugConfiguration(_folder: vscode.WorkspaceFolder | undefined, debugConfiguration: ResolvedDebugConfiguration): vscode.DebugConfiguration {
-            if (debugConfiguration && debugConfiguration.type && debugConfiguration.dockerOptions && debugConfiguration.dockerOptions.dockerServerReadyAction) {
-                if (!trackers.has(debugConfiguration.type)) {
-                    trackers.add(debugConfiguration.type);
-                    startTrackerForType(context, debugConfiguration.type);
+        if (!detector) {
+            return;
+        }
+
+        try {
+            const container = ext.dockerode.getContainer(containerName);
+
+            container.logs(
+                {
+                    follow: true,
+                    stdout: true
+                },
+                (err, stream) => {
+                    if (err) {
+                        return;
+                    }
+
+                    this.logStream = <LogStream>stream;
+                    this.logStream.on('data', (data) => {
+                        // tslint:disable-next-line:no-unsafe-any
+                        detector.detectPattern(data.toString());
+                    });
+                });
+        } catch { }
+    }
+
+    private disposeStream() : void {
```

Comment:
Nit: Since this is only called from one place (in `super()`), the implementation could just be moved there.